services/wfm/src/main/java/org/openkilda/wfm/topology/flowhs/service/FlowRerouteService.java

```diff
@@ -151,12 +156,54 @@ public void handleTimeout(String key) {
         removeIfFinished(fsm, key);
     }
 
+    private void initReroute(FlowRerouteFact reroute) {
+        final CommandContext commandContext = reroute.getCommandContext();
+
+        String eventKey = commandContext.getCorrelationId();
+        if (flowEventRepository.existsByTaskId(eventKey)) {
+            log.error("Attempt to reuse key {}, but there's a history record(s) for it.", eventKey);
+            return;
+        }
+
+        final String flowId =  reroute.getFlowId();
+        final String key = reroute.getKey();
+        FlowRerouteFsm fsm = fsmFactory.newInstance(commandContext, flowId);
+        fsms.put(key, fsm);
+
+        FlowRerouteContext context = FlowRerouteContext.builder()
+                .flowId(flowId)
+                .pathsToReroute(reroute.getPathsToReroute())
+                .forceReroute(reroute.isForceReroute())
+                .rerouteReason(reroute.getRerouteReason())
+                .build();
+        fsmExecutor.fire(fsm, Event.NEXT, context);
+
+        removeIfFinished(fsm, key);
+    }
+
     private void removeIfFinished(FlowRerouteFsm fsm, String key) {
         if (fsm.isTerminated()) {
-            log.debug("FSM with key {} is finished with state {}", key, fsm.getCurrentState());
-            fsms.remove(key);
+            performHousekeeping(fsm, key);
 
-            carrier.cancelTimeoutCallback(key);
+            // use some sort of recursion here, because iterative way require too complex scheme to clean/use retryQueue
+            retryQueue.peek(fsm.getFlowId()).ifPresent(carrier::injectRetry);
+        }
+    }
+
+    private void performHousekeeping(FlowRerouteFsm fsm, String key) {
+        log.debug("FSM with key {} is finished with state {}", key, fsm.getCurrentState());
+        fsms.remove(key);
+
+        carrier.cancelTimeoutCallback(key);
+
+        FlowRerouteFact reroute = retryQueue.remove(fsm.getFlowId())
+                .orElseThrow(() -> new IllegalStateException(String.format(
```

Comment:
will this exception be raised each time there are no pending reroutes in a queue?